test/unit/bigip/test_bigip_software_facts.py

```diff
@@ -0,0 +1,389 @@
+        )
+
+        with pytest.raises(F5ModuleError) as err:
+            msg = '"foo" is not a supported filter. Supported key values are: name, build, version, status, active'
+            FactManagerBase(client)
+
+        assert err.value.message == msg
+
+    def test_invalid_filter_format_raises(self, *args):
+        set_module_args(dict(
+            server='localhost',
+            password='password',
+            user='admin',
+            filter='foobar'
+        ))
+
+        client = AnsibleF5Client(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            f5_product_name=self.spec.f5_product_name
+        )
+
+        with pytest.raises(F5ModuleError) as err:
+            msg = '"foobar" is not a valid filter format. Filters must have key:value format'
+            FactManagerBase(client)
+
+        assert err.value.message == msg
+
+    def test_invalid_filter_empty_key_raises(self, *args):
+        set_module_args(dict(
+            server='localhost',
+            password='password',
+            user='admin',
+            filter=':foobar'
+        ))
+
+        client = AnsibleF5Client(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            f5_product_name=self.spec.f5_product_name
+        )
+
+        with pytest.raises(F5ModuleError) as err:
+            msg = '":foobar" is not a valid filter format. Filters must have key:value format'
+            FactManagerBase(client)
+
+        assert err.value.message == msg
+
+    def test_invalid_filter_empty_value_raises(self, *args):
+        set_module_args(dict(
+            server='localhost',
+            password='password',
+            user='admin',
+            filter='name:'
+        ))
+
+        client = AnsibleF5Client(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            f5_product_name=self.spec.f5_product_name
+        )
+
+        with pytest.raises(F5ModuleError) as err:
+            msg = '"name:" is not a valid filter format. Filters must have key:value format'
+            FactManagerBase(client)
+
+        assert err.value.message == msg
+
+    def test_invalid_include_raises(self, *args):
+        set_module_args(dict(
+            server='localhost',
+            password='password',
+            user='admin',
+            include=['bar', 'foo']
+        ))
+
+        client = AnsibleF5Client(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            f5_product_name=self.spec.f5_product_name
+        )
+
+        with pytest.raises(F5ModuleError) as err:
+            msg = 'Include parameter may only be specified as one or more of the following: all, volume, image, hotfix'
+            FactManagerBase(client)
+
+        assert err.value.message == msg
```

Comment:
to be python 3 compatible, all of these err.value.message assertions need to be `str(err.value)`